### DISCUSSION TITLE
Inverting PNG Files

### DISCUSSION THREAD

**Discussion by Yodazon** [python-pillow/Pillow#7473] [Q&A] [answered]
Hi everyone!
My goal is to invert some PNG files for an ML project I am doing. Currently I am using the ImageOps.invert module to achieve this.

I am running into an error stating "OSError: not supported for this image mode" for when I attempt this.
Earlier in the same project I utilize ImageOps.grayscale on the same images and it works like a charm. Therefore I am unsure where the issue lies.

Help is appreciated! Thanks

**Comment by bigcat88** [comment]
It will be helpful to first identify the version of Pillow and the example you are reproducing, where this error occurs.

**Reply by Yodazon** [reply]
The version of Pillow that I am using is 10.0.0.

The code is the following:

```python
from os import listdir
import os.path
import os
from PIL import Image as PImage
from PIL import ImageOps

def loadImages(path):
    imagesList = listdir(path)
    loadedImages = []
    for image in imagesList:
        img = PImage.open(path + image)
        loadedImages.append(img)
    return loadedImages

def grayScale(imgs):
    converted = []
    for img in imgs:
        gray = ImageOps.grayscale(img)
        converted.append(gray)
    return converted

def invertImage(imgs):
    converted = []
    for img in imgs:
        invert = ImageOps.invert(img)
        converted.append(invert)
    return converted
    
path = "my/path/here/"
imgs = loadImages(path)
grayScaleImgs = grayScale(imgs)
invertImgs = invertImage(imgs)
```
The difference between the two functions of 'grayScale' and 'invertImage' is the operation I would like performed.

**Reply by Yay295** [reply]
try this:
```python
from os import listdir
from PIL import Image, ImageOps

path = "my/path/here/"

for im_name in listdir(path):
    with Image.open(path + im_name) as img:
        print(path + im_name, 'has image mode', img.mode)
        ImageOps.invert(img)
```
This should print the mode of each image before trying to invert it, so you'll be able to see what mode it's failing on.

**Reply by Yodazon** [reply]
Looks like the mode that it fails on is the RGBA image mode

**Reply by radarhere** [reply]
Thanks. I'm guessing that you would like to invert the red, green and blue channels, but keep the alpha channel the same. If so, try the following
```python
def invertImage(imgs):
    converted = []
    for img in imgs:
        if img.mode == "RGBA":
            img_a = img.getchannel("A")
            img_rgb = img.convert("RGB")
            invert = ImageOps.invert(img_rgb)
            invert.putalpha(img_a)
        else:
            invert = ImageOps.invert(img)
        converted.append(invert)
    return converted
```

**Reply by Yodazon** [reply]
I tried out the cost you posted @radarhere and I am given the same final error. Full error message is as followed:
```pytb
 ImageOps.invert(img)
  File "my\path\here\Python\Python311\Lib\site-packages\PIL\ImageOps.py", line 534, in invert
    return image.point(lut) if image.mode == "1" else _lut(image, lut)
                                                      ^^^^^^^^^^^^^^^^
  File "my\path\here\Python\Python311\Lib\site-packages\PIL\ImageOps.py", line 60, in _lut
    raise OSError(msg)
OSError: not supported for this image mode
```

**Reply by radarhere** [reply, marked as answer]
The error is being raised on `ImageOps.invert(img)`, which is not in the "RGBA" part of my code - this makes me think that I've solved your RGBA problem, but there is another mode you are having trouble with as well.

Does this code print out something like "Image mode: RGBA" just before the traceback? What is that mode?

```python
def invertImage(imgs):
    converted = []
    for img in imgs:
        if img.mode == "RGBA":
            img_a = img.getchannel("A")
            img_rgb = img.convert("RGB")
            invert = ImageOps.invert(img_rgb)
            invert.putalpha(img_a)
        else:
            try:
                invert = ImageOps.invert(img)
            except OSError as e:
                print("Image mode:", img.mode)
                raise e
        converted.append(invert)
    return converted
```

**Reply by radarhere** [reply]
@Yodazon is there any further help you need?

**Reply by Yodazon** [reply]
Hi there, 

I ran the code in the previous comment and still had the same issue. Output is as followed:
```
my/path/here/ Photo.svg.png has image mode RGBA
Traceback (most recent call last):
  File "my/path/here/", line 80, in <module>
    ImageOps.invert(img)
  File "my\path\here\Python\Python311\Lib\site-packages\PIL\ImageOps.py", line 534, in invert
    return image.point(lut) if image.mode == "1" else _lut(image, lut)
                                                      ^^^^^^^^^^^^^^^^
  File "my\path\here\Python\Python311\Lib\site-packages\PIL\ImageOps.py", line 60, in _lut
    raise OSError(msg)
OSError: not supported for this image mode

```

**Reply by radarhere** [reply]
`ImageOps.invert(img)` is not a line in my last comment with code. `invert = ImageOps.invert(img)` is the closest line. So it doesn't look like you're running the code that I suggested.

**Reply by Yodazon** [reply]
Yes, I looked into it and you are correct! It was some of the test code from earlier. The solution you posted works! I am curious as to why it works though

**Reply by radarhere** [reply]
There shouldn't be any functional difference between my [first code example](https://github.com/python-pillow/Pillow/discussions/7473#discussioncomment-7308018) and my [second code example.](https://github.com/python-pillow/Pillow/discussions/7473#discussioncomment-7309134) The only difference should be that the second one prints the image mode when an error occurs.
